python/revng/scripting/projects/project.py

```diff
@@ -0,0 +1,245 @@
+        """
+        pass
+
+    def get_artifacts(self, params: Dict[str, List[StructBase]]) -> Dict[str, str | bytes]:
+        """
+        Allows fetching multiple artifacts at once. The `params` is a dict containing
+        the name of the artifact and a list of targets (it can be empty to fetch all the
+        targets).
+        Example `params`:
+        params = {
+            "disassemble": [Function_1, Function_2],
+            "decompile": []
+        }
+        """
+        result: Dict[str, str | bytes] = {}
+        for artifact_name, targets in params.items():
+            result[artifact_name] = self._get_artifact(artifact_name, targets)
+
+        return result
+
+    @abstractmethod
+    def analyze(
+        self, analysis_name: str, targets: Dict[str, List[str]] = {}, options: Dict[str, str] = {}
+    ):
+        """
+        Run a single analysis.
+        """
+        pass
+
+    @abstractmethod
+    def analyses_list(self, analysis_name: str):
+        """
+        Run analysis list, these are predefined list of analysis that run sequentially.
+        """
+        pass
+
+    @abstractmethod
+    def commit(self):
+        """
+        Persist the changes from `self.model` to the backend.
+        """
+        pass
+
+    def revert(self):
+        """
+        Revert changes made to the `self.model` since the last call to `commit()`.
+        """
+        self._set_model(deepcopy(self.last_saved_model))
+
+    @abstractmethod
+    def _get_pipeline_description(self):
+        """
+        Define how to implement getting the pipeline description data, The method should parse the
+        data as `yaml` and store it in the variable `self.pipeline_description`
+        """
+        pass
+
+    def import_and_analyze(self, input_binary_path: str):
+        """
+        A helper method for setting up the project. This method imports
+        the binary, runs the `revng` and `revng-c` initial auto analysis,
+        gets the pipeline description and sets the mixins.
+        """
+        self.import_binary(input_binary_path)
+        self.analyses_list("revng-initial-auto-analysis")
+        self.analyses_list("revng-c-initial-auto-analysis")
+        self._get_pipeline_description()
+        self._set_model_mixins()
+
+    def get_targets(self, analysis_name: str, targets: List[Function]) -> Dict[str, List[str]]:
+        """
+        Create a dict of targets from the container input and the list `Function` objects
+        """
+        container_input = self._get_analysis_inputs(analysis_name)
+
+        _targets: Dict[str, List[str]] = {}
+        for input, kinds in container_input.items():  # noqa: A001
+            _targets[input] = []
+            for kind in kinds:
+                if kind == "root":
+                    _targets[input].append(f":{kind}")
+                else:
+                    _targets[input].extend(f"{target.key()}:{kind}" for target in targets)
```

Comment:
Make this kind-agnostic

Reply:
Should I remove this method altogether? I don't really see the point in having it if its only looping the `targets` and getting their `.key()`. Then I could make the method `_get_analysis_inputs()` public and let the user assemble the targets. Or how did you envision this method?

Reply:
Hmmm, I'd drop this function for the time being (also reword the passage in docs). There needs to be a facility like this, but this is not in scope for now.

Reply:
I removed the function and added the functionality to the `test` and the `docs`